include/pstl_offload/internal/usm_memory_replacement.h

```diff
@@ -53,34 +53,58 @@ __get_offload_device()
 class __offload_policy_holder_type
 {
     using __set_active_device_func_type = void (*)(sycl::device*);
+    using __get_offload_device_func_type = sycl::device (*)();
 
   public:
     // Since the global object of __offload_policy_holder_type is static but the constructor
     // of the class is inline, we need to avoid calling static functions inside of the constructor
     // and pass the pointer to exact function as an argument to guarantee that the correct __active_device
     // would be stored in each translation unit
-    __offload_policy_holder_type(sycl::device&& __offload_device,
+    __offload_policy_holder_type(__get_offload_device_func_type __get_offload_device_func,
                                  __set_active_device_func_type __set_active_device_func)
-        : _M_offload_device(std::move(__offload_device)), _M_offload_policy(_M_offload_device),
-          _M_set_active_device(__set_active_device_func)
+        : _M_set_active_device(__set_active_device_func)
     {
-        _M_set_active_device(&_M_offload_device);
+        try
+        {
+            _M_offload_device = __get_offload_device_func();
+        }
+        catch (const sycl::exception& e)
+        {
+            // e.code() == sycl::errc::runtime means occurs when unable to get offload device. Do not pass
+            // an exception, as ctor is called for a static object and the exception can't be processed.
+            // Remember the situation and re-throw exception when asked for the policy from user's code.
+            if (e.code() != sycl::errc::runtime)
```

Comment:
What kind of other exceptions can be thrown when constructing the device? 
If we are re-throwing all exceptions other then `sycl::errc::runtime`, it seems like there would be an other exception thrown by the constructor of the global object and hence it cannot be catched by the user.

Reply:
> If we are re-throwing all exceptions other then sycl::errc::runtime, it seems like there would be an other exception thrown by the constructor of the global object and hence it cannot be catched by the user.

Exactly, and that's the plan. We are not going to mitigate situation "something bad occurs, let's hide it for a while", but only cover the expected situation.

Possible alternative is `std::exception_ptr`, `std::rethrow_exception()` etc. I don't like it, because it's non-trivial processing of something unknown.

Sure, this might be discussed.